Use $scope observables for dependency-based data flow

uniphil opened the pull request.

uniphil committed:
Merge remote-tracking branch 'origin/edudash-2.0' into fast-markers

uniphil committed:
Simplify feature properties structure

uniphil committed:
$scope API cleanup: feature hover

uniphil committed:
$scope API cleanup: schoolType

uniphil committed:
Refactor: move map UI functions together

uniphil committed:
Use the scope methods and watchers

uniphil committed:
Fix view mode setting

stragglers changed to the new setViewMode method

uniphil committed:
Refactor: begin cleaning up setSchool -> select

uniphil committed:
Refactor: data loading flow for points from ckan

uniphil committed:
Data flow for colouring pins

pt_ratio is missing in CKAN data?

uniphil committed:
Remove some more cartodb remnants

uniphil committed:
fix jquery dependency error

I think select2 used to get jquery from cartodb (?!). Now
it is loaded after jquery, so it works again

uniphil committed:
add some safety and log fetch errors

uniphil committed:
Track schools as their data, not as leaflet layers

specifically for hovered, selected, etc.

uniphil committed:
Migrate search to new data flow

uniphil committed:
reindent

uniphil committed:
Use local data for secondary improved list

uniphil committed:
use moreThan40 toggle

uniphil committed:
Re-hack the hack to make fix layer reloading

uniphil committed:
Make primary schools work again

uniphil committed:
Constant-time school lookup by ID

Doesn't really need to be a promise any more... could be a small performance
win to remove that later

uniphil committed:
Make the year selector work again

Fixed markup also, which probably broke the styles

uniphil committed:
Merge branch 'edudash-2.0' into refactor

uniphil committed:
fix indentation

uniphil committed:
Leave school properties as database column names

More shouty, but simpler and more consistent

uniphil committed:
Fix secondary ranked list

uniphil committed:
Rank primary schools

uniphil committed:
remove a stray log

uniphil committed:
Make the guages work again

still need to move them to the new data flow

uniphil committed:
Keep year aggregates in new data flow for pass rate

uniphil committed:
Restore pass rate over time chart on new data flow

uniphil committed:
Fix indentation and add a safety check

uniphil committed:
Rename more school properties to match the db fields

uniphil committed:
reindent

uniphil committed:
Hook up pass-rate-time chart with new data flow

uniphil committed:
respacing

uniphil committed:
Reindent

uniphil committed:
Use ranks from new data flow

uniphil committed:
Trim the scope a bit

Move selected-school-related async data onto the school object -- angular
apparently does a deep check of objects, so it works.

uniphil committed:
Fix passrate over time for primary schools

uniphil committed:
Cleanup time

remove some dead code and legacy stuff

esoliani commented on app/scripts/services/opendataapi.coffee:
```diff
@@ -23,6 +23,29 @@ angular.module 'edudashAppSrv'
           'improvement': 'bba2cbbb-97fb-48b1-aa51-8db69279fbc5'
         secondary: '743e5062-54ae-4c96-a826-16151b6f636b'
 
+      converters =
```
Nice way to parse ckan data

esoliani commented on app/views/ratepasschart.html:
```diff
@@ -1,17 +1,21 @@
+    <div ng-class="{'col-xs-3': selectedSchool.PASS_RATE == '100', 'col-xs-2': selectedSchool.PASS_RATE != '100'}">
+      <p class="percentNumber widgetnumber">{{ selectedSchool.PASS_RATE| number:0}}%</p>
     </div>
-    <div class="row" ng-if="selectedSchool.PASS_RATE == '0'">
```
why do you remove the NA text? the condition should changed

esoliani commented on app/views/ratepasschart.html:
```diff
@@ -1,17 +1,21 @@
+          ng-repeat="i in getTimes(10-(selectedSchool.PASS_RATE/10)) track by $index"></div>
     </div>
+  </div>
+  <div class="row" ng-if="selectedSchool.PASS_RATE == '0'">
```
does this condition works? since the pass_rate is numeric now

uniphil replied:
haha it does, but it's terrible and thanks for catching it. `==` will do implicit conversions in JS. I'll swap it out for `===` and make it a literal zero

esoliani commented on app/views/schoollist.html:
```diff
@@ -1,17 +1,26 @@
+        <strong>{{ school.NAME }}</strong>
+      </td>
+      <td ng-class="{
+        'passrategreen': school[property]>=(max+0),
```
I think we could remove this conversion

uniphil committed:
Load extended shool properties dynamically

Reduces the main payload from 4MB to 2.9MB (before GZIP) on
page-load.

Through `cors-anywhere` on my connection, this reduces the wait
from 6.2s to 5.0s on average

uniphil committed:
Make implicit comparissons strict

Use === and !== instead of == and !=-, and change the
string numebers to literals

uniphil committed:
Make implicit type conversions explicit

Strinct comparissons, remove unneeded explicit conversions

uniphil committed:
Async load all school ranks on pageload

Reverts the change to load school ranks dynamically when a school is
selected, but adds a separate async request to get the rankings.

the spinner blocks the UI for the second async request, so it should be
changed to cover only the components whic depend on pending requests

ddimunzio commented:
let's merge and fix issues afterwards
